src/test/java/com/aws/greengrass/mqttbridge/util/BatchedSubscriberTest.java

```diff
@@ -0,0 +1,257 @@
+
+            numTimesCalled.getAndIncrement();
+            testComplete.countDown();
+        });
+        bs.subscribe();
+
+        try {
+            waitForChangesToQueue(topics, () -> IntStream.range(0, 5).forEach(i ->
+                    topics.updateFromMap(Utils.immutableMap("key", i), mergeBehavior.get())
+            ));
+            assertTrue(testComplete.await(5L, TimeUnit.SECONDS));
+            topics.context.waitForPublishQueueToClear();
+        } finally {
+            bs.unsubscribe();
+            topics.context.close();
+        }
+
+        assertEquals(1, numInitializations.get());
+        assertEquals(1, numTimesCalled.get());
+    }
+
+    @Test
+    void GIVEN_subscribe_to_topics_WHEN_separate_events_THEN_callback_runs_every_time() throws Exception {
+        Topics topics = Topics.of(new Context(), "topic", null);
+
+        final int expectedNumChanges = 10;
+
+        AtomicInteger numInitializations = new AtomicInteger();
+        AtomicInteger numTimesCalled = new AtomicInteger();
+        CountDownLatch testComplete = new CountDownLatch(1);
+
+        // Lookup the topic which will be created by the updateMap call below
+        // so that this event has time to propagate and then we have a known number of updates for below
+        topics.lookup("key");
+        topics.context.waitForPublishQueueToClear();
+
+        BatchedSubscriber bs = new BatchedSubscriber(topics, (what, whatChanged) -> {
+            if (what == WhatHappened.initialized) {
+                numInitializations.incrementAndGet();
+                return;
+            }
+
+            if (numTimesCalled.incrementAndGet() >= expectedNumChanges) {
+                testComplete.countDown();
+            }
+        });
+        bs.subscribe();
+
+        try {
+            IntStream.range(0, expectedNumChanges).forEach(i -> {
+                topics.updateFromMap(Utils.immutableMap("key", i), mergeBehavior.get());
+                topics.context.waitForPublishQueueToClear();
+            });
+
+            assertTrue(testComplete.await(5L, TimeUnit.SECONDS));
+            topics.context.waitForPublishQueueToClear();
+        } finally {
+            bs.unsubscribe();
+            topics.context.close();
+        }
+
+        assertEquals(1, numInitializations.get());
+        assertEquals(expectedNumChanges, numTimesCalled.get());
+    }
+
+    /**
+     * For a consistent test scenario, we ensure all config changes
+     * are properly queued before batched subscriber does its work.
+     *
+     * @param node         topic or topics
+     * @param queueChanges action that modifies the provided topic(s)
+     */
+    private void waitForChangesToQueue(Node node, Runnable queueChanges) {
+        CountDownLatch waitForChangesToQueue = new CountDownLatch(1);
+        node.context.runOnPublishQueue(() -> {
+            try {
+                waitForChangesToQueue.await();
+            } catch (InterruptedException e) {
+                fail(e);
```

Comment:
calling fail here will _not_ cause the test to fail because this is running in a different thread. Only exceptions on the main thread will cause the test to fail. There is however, an "ExceptionLogProtector" which may kick in and fail the test for logged exceptions. In general though, make sure that any test expectations happen on the main thread.

Reply:
👍 fixed